engine/consensus/compliance/engine.go

```diff
@@ -546,43 +551,54 @@ func (e *Engine) onBlockVote(originID flow.Identifier, vote *messages.BlockVote)
 	return nil
 }
 
-// processPendingChildren checks if there are proposals connected to the given
+// processBlockAndDescendants processes  if there are proposals connected to the given
 // parent block that was just processed; if this is the case, they should now
 // all be validly connected to the finalized state and we should process them.
-func (e *Engine) processPendingChildren(header *flow.Header) error {
-	blockID := header.ID()
+func (e *Engine) processBlockAndDescendants(proposal *messages.BlockProposal) error {
+	blockID := proposal.Header.ID()
+
+	// process block itself
+	err := e.processBlockProposal(proposal)
+	if err != nil {
+		return fmt.Errorf("failed to process block %x: %w", blockID, err)
+	}
 
-	// check if there are any children for this parent in the cache
+	// process all children
+	// do not break on on invalid or outdated blocks as they should not prevent us
+	// from processing other valid children
```

Comment:
I think it would make sense to order children from the oldest to the youngest here before processing. In that case, it does indeed make sense to stop processing once we reach an invalid (drop descendants) or unverifiable (resume next time) child.

Reply:
My understanding is that `ByParentID` returns only direct children of the block (with the same height = proposal + 1), not a list of connected descendants.
The recursion takes care of travelling the chain of descendants from one child to another.